src/ring_theory/derivations.lean

```diff
@@ -0,0 +1,200 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Nicolò Cavalleri.
+-/
+
+import algebra.lie_algebra
+import tactic
+
+/-!
+# Derivations
+
+This file defines derivation.
+
+IMPORTANT: this file is just a stub to go on with some PRs in the geometry section. It only
+implements the definition of derivations in commutative algebra. This will soon change: as soon
+as bimodules will be there in mathlib I will change this file to take into account the
+non-commutative case. Any development on the theory of derivations is discouraged until the
+definitive definition of derivation will be implemented.
+-/
+
+open algebra ring_hom
+
+/-- Natural way to define a scalar multiplication over a ring `R` for a module `M` over an
+algebra `A` over `R`. -/
+def transitive_scalar (R : Type*) (A : Type*) (M : Type*)
+  [comm_semiring R] [semiring A] [algebra R A]
+  [add_comm_monoid M] [semimodule A M] : has_scalar R M :=
+{ smul := λ r m, ((algebra_map R A) r) • m, }
+
+/-- With the operation defined above the module `M` is also naturally a module over `R`. -/
+def transitive_module (R : Type*) (A : Type*) (M : Type*)
+  [comm_semiring R] [semiring A] [algebra R A] [add_comm_monoid M] [semimodule A M] :
+  semimodule R M :=
+{ smul_add := λ r x y, smul_add _ _ _,
+  smul_zero := λ r, smul_zero _,
+  zero_smul := λ x, show algebra_map R A 0 • x = 0, by rw [map_zero, zero_smul],
+  one_smul := λ x, show algebra_map R A 1 • x = x, by rw [map_one, one_smul],
+  mul_smul := λ r s x, show algebra_map R A (r * s) • x =
+    algebra_map R A r • algebra_map R A s • x, by rw [map_mul, mul_smul],
+  add_smul := λ r s x, show algebra_map R A (r + s) • x =
+    algebra_map R A r • x + algebra_map R A s • x, by rw [map_add, add_smul],
+  .. transitive_scalar R A M }
+
+/-- This class register an instance that the scalar multiplication over `M` is indeed defined as
+above in `transitive_scalar`. -/
+@[protect_proj]
+class compatible_semimodule (R : Type*) (A : Type*) [comm_semiring R] [semiring A] [algebra R A]
+(M : Type*) [add_comm_monoid M] [semimodule A M] [semimodule R M] :=
+(compatible_smul (r : R) (m : M) : r • m = ((algebra_map R A) r) • m)
```

Comment:
Could you try making `is_algebra_tower` extend this class?

Reply:
I am not really sure where and how `is_algebra_tower` is used. I do not really know this section of the library. I can try to do it but I am scared I'd make a mess. Please feel free to push a commit to this branch if you really want this in this PR!

Reply:
I understand. I hope to have some time to look at that tomorrow.

Reply:
Ok I did replace it with `is_scalar_tower`